src/controllers/midi/midicontroller.cpp

```diff
@@ -30,22 +30,29 @@ QString MidiController::mappingExtension() {
     return MIDI_MAPPING_EXTENSION;
 }
 
-void MidiController::visit(const LegacyMidiControllerMapping* mapping) {
-    m_mapping = *mapping;
-    emit mappingLoaded(getMapping());
+void MidiController::setMapping(std::shared_ptr<LegacyControllerMapping> pMapping) {
```

Comment:
Move this code into a static, generic function of the base class?

```
template<typename T>
static std::shared_ptr<T> downcastAndTakeOwnership(std::shared_ptr<LegacyControllerMapping>)`
```

It is already repeated multiple times now.

Reply:
I pushed a WIP commit which fails to link. Any hints? I tried removing `static`.

Reply:
ah, I had to move the implementation to the header file... :upside_down_face:

Reply:
Of course, sure, didn't notice. The generic implementation has to be placed in the header file. But in contrast to regular functions you don't have to use the keyword `inline` to prevent multiple implementations when included. Inconsistent and easy to confuse.